src/java.base/share/classes/jdk/internal/foreign/LayoutRecordMapper.java

```diff
@@ -0,0 +1,641 @@
+                            } catch (NoSuchMethodException | IllegalAccessException e) {
+                                throw new InternalError(e);
+                            }
+                        }
+                        case GroupLayout gl -> {
+                            @SuppressWarnings("unchecked")
+                            var componentType = (Class<? extends Record>) cl.component().getType();
+                            var componentMapper = recordMapper(componentType, gl, byteOffset);
+                            try {
+                                var mt = MethodType.methodType(Record.class, MemorySegment.class);
+                                var mh = LOOKUP.findVirtual(LayoutRecordMapper.class, "apply", mt);
+                                // (LayoutRecordAccessor, MemorySegment)Record -> (MemorySegment)Record
+                                mh = MethodHandles.insertArguments(mh, 0, componentMapper);
+                                // (MemorySegment)Record -> (MemorySegment)componentType
+                                yield MethodHandles.explicitCastArguments(mh, MethodType.methodType(componentType, MemorySegment.class));
+                            } catch (NoSuchMethodException | IllegalAccessException e) {
+                                throw new RuntimeException(e);
+                            }
+                        }
+                        case SequenceLayout sl -> {
+                            try {
+                                var componentType = cl.component().getType();
+                                if (!componentType.isArray()) {
+                                    throw new IllegalArgumentException("Unable to map '" + sl +
+                                            "' because the component '" + componentType.getName() + " " + name + "' is not an array");
+                                }
+
+                                MultidimensionalSequenceLayoutInfo info = MultidimensionalSequenceLayoutInfo.of(sl);
+
+                                if (info.elementLayout() instanceof ValueLayout.OfBoolean) {
+                                    throw new IllegalArgumentException("Arrays of booleans (" + info.elementLayout() + ") are not supported");
+                                }
+
+                                if (dimensionOf(componentType) != info.sequences().size()) {
+                                    throw new IllegalArgumentException("Unable to map '" + sl + "'" +
+                                            " of dimension " + info.sequences().size() +
+                                            " because the component '" + componentType.getName() + " " + name + "'" +
+                                            " has a dimension of " + dimensionOf(componentType));
+                                }
+
+                                // Handle multi-dimensional arrays
+                                if (info.sequences().size() > 1) {
```

Comment:
I have a feeling that this could be simplified if we embraced recursion, and defined the mapping as a function taking three arguments: offset, layout and carrier. The output of that function is a MethodHandle. Record mapper is not special - I mean, it is special as that's the only entry point we provide in the API, but that is based on a more general way to map layouts and carriers to value, so we should probably write that general function first, and then map that to the recordMapper API point.

Reply:
> I have a feeling that this could be simplified if we embraced recursion, and defined the mapping as a function taking three arguments: offset, layout and carrier. The output of that function is a MethodHandle. Record mapper is not special - I mean, it is special as that's the only entry point we provide in the API, but that is based on a more general way to map layouts and carriers to value, so we should probably write that general function first, and then map that to the recordMapper API point.

That said, I think your optimization of the one-dimensional case also makes sense, as in those cases we know we can just bulk copy into the target array. So perhaps, the split between one-dimensional and multi-dimensional is gonna remain (but I think in terms of the javadoc, the split shouldn't be there).